relay_address_generator_range.go

```diff
@@ -0,0 +1,105 @@
+package turn
+
+import (
+	"errors"
+	"fmt"
+	"github.com/pion/transport/vnet"
+	"math/rand"
+	"net"
+	"strings"
+	"sync"
+	"time"
+)
+
+// RelayAddressGeneratorPortRange can be used to only allocate connections inside a defined port range.
+// Similar to the RelayAddressGeneratorStatic a static ip address can be set.
+type RelayAddressGeneratorPortRange struct {
+	// RelayAddress is the IP returned to the user when the relay is created
+	RelayAddress net.IP
+
+	// MinPort the minimum port to allocate
+	MinPort uint16
+	// MaxPort the maximum (inclusive) port to allocate
+	MaxPort uint16
+
+	// MaxRetries the amount of tries to allocate a random port in the defined range
+	MaxRetries int
+
+	// Rand the random source of numbers
+	Rand      *rand.Rand
+	randMutex sync.Mutex
+
+	// Address is passed to Listen/ListenPacket when creating the Relay
+	Address string
+
+	Net *vnet.Net
+}
+
+// Validate is called on server startup and confirms the RelayAddressGenerator is properly configured
+func (r *RelayAddressGeneratorPortRange) Validate() error {
+	if r.Net == nil {
+		r.Net = vnet.NewNet(nil)
+	}
+
+	if r.Rand == nil {
+		r.Rand = rand.New(rand.NewSource(time.Now().UnixNano()))
+	}
+
+	if r.MaxRetries == 0 {
+		r.MaxRetries = 10
+	}
+
+	switch {
+	case r.MinPort == 0:
+		return errors.New("MinPort must not be 0")
+	case r.MaxPort == 0:
+		return errors.New("MaxPort must not be 0")
+	case r.RelayAddress == nil:
+		return errRelayAddressInvalid
+	case r.Address == "":
+		return errListeningAddressInvalid
+	default:
+		return nil
+	}
+}
+
+func (r *RelayAddressGeneratorPortRange) randPort() uint16 {
+	r.randMutex.Lock()
+	defer r.randMutex.Unlock()
+	return r.MinPort + uint16(r.Rand.Intn(int((r.MaxPort+1)-r.MinPort)))
+}
+
+// AllocatePacketConn generates a new PacketConn to receive traffic on and the IP/Port to populate the allocation response with
+func (r *RelayAddressGeneratorPortRange) AllocatePacketConn(network string, requestedPort int) (net.PacketConn, net.Addr, error) {
+	if requestedPort != 0 {
+		conn, err := r.Net.ListenPacket(network, fmt.Sprintf("%s:%d", r.Address, requestedPort))
+		if err != nil {
+			return nil, nil, err
+		}
+		relayAddr := conn.LocalAddr().(*net.UDPAddr)
+		relayAddr.IP = r.RelayAddress
+		return conn, relayAddr, nil
+	}
+
+	for try := 0; try < r.MaxRetries; try++ {
+		conn, err := r.Net.ListenPacket(network, fmt.Sprintf("%s:%d", r.Address, r.randPort()))
+
+		if err != nil {
+			if strings.Contains(err.Error(), "address already in use") {
```

Comment:
Is this okay? it seems a little hacky to me but I couldn't find a better solution.

Reply:
I would drop the `string.Contains`. Might be worth logging the error though!

Reply:
But we need the check (or something similar). Otherwise, AllocatePacketConn will err and not retry if the rand source generates an already binded port.

Reply:
Sorry I rushed my comment :( I would do this. Don't worry about catching the specific error, since it isn't typed I could see this causing issues on platforms we don't test for.

```
if err != nil {
  continue
}
```

Reply:
If you agree with that I would love to merge (and tag) this today! Happy to make the change myself and squash + push (with you as the author!)

Reply:
Yeah, this is fine by me :+1:.